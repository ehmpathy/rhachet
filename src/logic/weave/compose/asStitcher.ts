import { Stitcher, GStitcher } from '../../../domain/objects/Stitcher';
import { Threads } from '../../../domain/objects/Threads';

/**
 * .what = applies a standardized type wrapper around a stitcher to enforce single-thread compatibility
 * .note
 *   - forces Threads<,single>, since by definition generic stitchers are always single threaded per role
 * .why =
 *   - enables composition of stitchers
 *   - without this, they're not composable, due to typescript(generic invariance on classes cause)
 *   - infers the type deeply
 */

export const asStitcher = <T extends Stitcher<GStitcher<any, any, any>>>(
  input: T,
): T extends Stitcher<GStitcher<infer TThreads, infer TContext, infer TOutput>>
  ? Stitcher<
      GStitcher<
        Threads<
          TThreads extends Threads<infer TContextDict, any>
            ? TContextDict
            : never,
          'single'
        >,
        TContext,
        TOutput
      >
    >
  : never => input as any;
