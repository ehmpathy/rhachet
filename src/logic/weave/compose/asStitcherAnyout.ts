import { Threads } from '../../../domain/objects';
import { GStitcher, Stitcher } from '../../../domain/objects/Stitcher';

/**
 * .what = safely casts a stitcher to preserve threads + context, but treat output as `any`
 * .why =
 *   - allows reuse of stitchers when only the input shape matters
 *   - ignores the output type, making it flexible for routes that discard outputs
 *   - avoids mismatches when flattening or combining steps
 */
export const asStitcherAnyout = <
  TDesired extends GStitcher<Threads<any>, any, any>,
>(
  input: Stitcher<GStitcher<TDesired['threads'], TDesired['context'], any>>,
): Stitcher<GStitcher<TDesired['threads'], TDesired['context'], any>> => input;
