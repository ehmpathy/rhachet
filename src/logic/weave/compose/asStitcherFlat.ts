import { GStitcher, Stitcher } from '../../../domain/objects/Stitcher';
import { Threads } from '../../../domain/objects/Threads';

/**
 * .what = narrows a stitcher with deeply inferred or recursive types into a pre-flattened form
 * .why =
 *   - enables you to simplify the types youre looking at, against what you expect
 *   - avoids TS2589 (excessively deep type instantiation), by memoization of the shape, tested against the inferred types of the stitcher so far
 *   - ensures downstream use of `asStitcher(...)` stays composable
 */
export const asStitcherFlat = <
  Flat extends GStitcher<Threads<any, 'single'>, any, any>,
>(
  input: Stitcher<Flat>,
): Stitcher<Flat> => input;
