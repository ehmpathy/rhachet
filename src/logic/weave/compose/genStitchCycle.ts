import { UniDuration } from '@ehmpathy/uni-time';
import { PickAny } from 'type-fns';

import { StitchCycle } from '../../../domain/objects/StitchCycle';
import {
  GStitcher,
  GStitcherFlat,
  Stitcher,
} from '../../../domain/objects/Stitcher';
import { Threads } from '../../../domain/objects/Threads';
import { GStitcherInferredFromCycle } from './GStitcherInferredFromCycle.generic';

/**
 * .what = generates a stitcher that repeats a step until the decider returns 'release' or 'halt'
 * .why =
 *   - supports runtime loops with fine-grained flow control
 * .note =
 *   - decider must return { choice: 'repeat' | 'release' | 'halt' }
 *   - halter is optional; defaults applied internally to prevent infinite loops
 */
export const genStitchCycle = <
  TRepeatee extends Stitcher<GStitcher<Threads<any, 'single'>, any, any>>,
  TDecider extends Stitcher<
    GStitcher<
      Threads<any, 'single'>,
      any,
      { choice: 'repeat' | 'release' | 'halt' }
    >
  >,
>(input: {
  slug: string;
  readme: string | null;
  repeatee: TRepeatee;
  decider: TDecider;
  halter?: {
    threshold: PickAny<{
      repetitions: number;
      duration: UniDuration;
    }>;
  };
}): StitchCycle<
  GStitcherFlat<GStitcherInferredFromCycle<TRepeatee, TDecider>>
> => {
  return new StitchCycle({
    form: 'CYCLE',
    slug: input.slug,
    readme: input.readme,
    repeatee: input.repeatee as any,
    decider: input.decider as any,
    halter: input.halter ?? {
      threshold: {
        // safe by default
        repetitions: 100,
        duration: { hours: 1 },
      },
    },
  });
};
