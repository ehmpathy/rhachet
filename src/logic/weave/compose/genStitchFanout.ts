import { Stitcher, GStitcher } from '../../../domain/objects/Stitcher';
import { GStitcherInferred } from './GStitcherInferred.generic';

/**
 * .what = generates a stitcher that fans out to multiple parallel stitchers, then concludes with a final one
 * .why =
 *   - enables parallel feedback or computation from multiple perspectives
 *   - merges results into a final output via the concluding stitcher
 */
export const genStitchFanout = <
  TParallels extends readonly [
    Stitcher<GStitcher<any, any, any>>,
    ...Stitcher<GStitcher<any, any, any>>[],
  ],
  TConclusion extends Stitcher<GStitcher<any, any, any>>,
>(input: {
  parallels: TParallels;
  conclusion: TConclusion;
}): Stitcher<GStitcherInferred<[...TParallels, TConclusion]>> => {
  return {
    form: 'ROUTE', // if youâ€™re using a special form, or otherwise mark as synthetic
    slug: `[fanout]:<${input.parallels.length}+1>`,
    sequence: [...input.parallels, input.conclusion] as const,
  } as unknown as Stitcher<GStitcherInferred<[...TParallels, TConclusion]>>;
};
