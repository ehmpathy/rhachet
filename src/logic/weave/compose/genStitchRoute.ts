import { StitchRoute } from '../../../domain/objects/StitchRoute';
import { GStitcher, Stitcher } from '../../../domain/objects/Stitcher';
import { Threads } from '../../../domain/objects/Threads';
import { GStitcherInferredFromRoute } from './GStitcherInferredFromRoute.generic';

/**
 * .what = generates a thoroughly typesafe stitch route
 * .why =
 *   - ensures typescript considers the input stitcher sequence to infer a composite stitcher for the stitch route declaration
 *   - results in a composite superset type of threads and context, plus accurate output type, on the output
 */
export const genStitchRoute = <
  TSequence extends readonly [
    // note: we only need to support single threaded here
    Stitcher<GStitcher<Threads<any, 'single'>, any, any>>,
    ...Stitcher<GStitcher<Threads<any, 'single'>, any, any>>[],
  ],
>(input: {
  slug: string;
  readme: string | null;
  sequence: TSequence;
}): StitchRoute<GStitcherInferredFromRoute<TSequence>> => {
  return new StitchRoute({
    form: 'ROUTE',
    slug: input.slug,
    readme: input.readme,
    sequence: input.sequence as any,
  });
};
