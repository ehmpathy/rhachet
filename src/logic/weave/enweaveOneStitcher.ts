import { UnexpectedCodePathError } from 'helpful-errors';

import { StitchSetEvent } from '../../domain/objects/StitchSetEvent';
import {
  GStitcher,
  Stitcher,
  StitcherForm,
} from '../../domain/objects/Stitcher';
import { enstitch } from '../stitch/enstitch';
import { withStitchTrail } from '../stitch/withStitchTrail';
import { enweaveOneFanout } from './enweaveOneFanout';
import { enweaveOneRoute } from './enweaveOneRoute';

/**
 * .what = a mechanism to enweave any stitcher
 * .why =
 *   - figures out which mechanism to call for each stitcher
 */
export const enweaveOneStitcher = withStitchTrail(
  async <TStitcher extends GStitcher>(
    input: {
      stitcher: Stitcher<TStitcher>;
      threads: TStitcher['threads'];
    },
    context: TStitcher['context'],
  ): Promise<StitchSetEvent<TStitcher['threads'], TStitcher['output']>> => {
    switch (input.stitcher.form) {
      // support step stitchers
      case StitcherForm.COMPUTE:
      case StitcherForm.IMAGINE:
        return enstitch(
          { stitcher: input.stitcher, threads: input.threads },
          context,
        );

      // support route stitcher
      case StitcherForm.ROUTE:
        return enweaveOneRoute(
          { stitcher: input.stitcher, threads: input.threads },
          context,
        );

      case StitcherForm.FANOUT:
        return enweaveOneFanout(
          { stitcher: input.stitcher, threads: input.threads },
          context,
        );
      // case StitcherForm.CHOICE:
      // case StitcherForm.CYCLE:
      // case StitcherForm.AWAIT:
      default:
        UnexpectedCodePathError.throw(
          'unsupported stitcher form in enweaveOneStitcher',
          {
            form: input.stitcher.form,
            stitcher: input.stitcher,
          },
        );
    }
  },
);
