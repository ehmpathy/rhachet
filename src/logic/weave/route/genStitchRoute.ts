import { StitchRoute } from '../../../domain/objects/StitchRoute';
import { GStitcher, Stitcher } from '../../../domain/objects/Stitcher';

type First<T extends any[]> = T extends [infer F, ...any[]] ? F : never;
type Last<T extends any[]> = T extends [...any[], infer L] ? L : never;

/**
 * .what = a mech to infer a GStitcher from a StitchRoute.Sequence
 */
type GStitcherInferred<TSequence extends Stitcher<GStitcher<any, any, any>>[]> =
  GStitcher<
    First<TSequence> extends Stitcher<infer G> ? G['threads'] : never,
    First<TSequence> extends Stitcher<infer G>
      ? G['procedure']['context']
      : never,
    Last<TSequence> extends Stitcher<infer G> ? G['output'] : never
  >;

export const genStitchRoute = <
  TSequence extends [
    Stitcher<GStitcher<any, any, any>>,
    ...Stitcher<GStitcher<any, any, any>>[],
  ],
>(input: {
  slug: string;
  name: string;
  description: string | null;
  sequence: TSequence;
}): StitchRoute<GStitcherInferred<TSequence>> => {
  return {
    slug: input.slug,
    name: input.name,
    description: input.description,
    sequence: input.sequence as any,
  };
};
