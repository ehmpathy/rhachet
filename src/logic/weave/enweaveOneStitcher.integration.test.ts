import { asUniDateTime, toMilliseconds } from '@ehmpathy/uni-time';
import { UnexpectedCodePathError } from 'helpful-errors';
import { given, then, when } from 'test-fns';
import { Empty } from 'type-fns';
import { getUuid } from 'uuid-fns';

import { genContextLogTrail } from '../../__test_assets__/genContextLogTrail';
import { genContextStitchTrail } from '../../__test_assets__/genContextStitchTrail';
import { getContextOpenAI } from '../../__test_assets__/getContextOpenAI';
import { genStitcherCodeFileRead } from '../../__test_assets__/stitchers/genStitcherCodeFileRead';
import { genStitcherCodeFileWrite } from '../../__test_assets__/stitchers/genStitcherCodeFileWrite';
import { genStitcherCodeReview } from '../../__test_assets__/stitchers/genStitcherCodeReviewImagine';
import { stitcherCodeDiffImagine } from '../../__test_assets__/stitchers/stitcherCodeDiffImagine';
import { getExampleThreadCodeArtist } from '../../__test_assets__/threads/codeArtist';
import { getExampleThreadCodeCritic } from '../../__test_assets__/threads/codeCritic';
import { exampleThreadDirector } from '../../__test_assets__/threads/director';
import { StitchStepCompute } from '../../domain/objects/StitchStep';
import { GStitcher } from '../../domain/objects/Stitcher';
import { Threads } from '../../domain/objects/Threads';
import { ContextOpenAI } from '../stitch/adapters/imagineViaOpenAI';
import { genThread } from '../thread/genThread';
import { asStitcher } from './compose/asStitcher';
import { asStitcherFlat } from './compose/asStitcherFlat';
import { genStitchCycle } from './compose/genStitchCycle';
import { genStitchFanout } from './compose/genStitchFanout';
import { genStitchRoute } from './compose/genStitchRoute';
import { enweaveOneStitcher } from './enweaveOneStitcher';

jest.setTimeout(toMilliseconds({ minutes: 3 }));

describe('enweaveOneStitcher', () => {
  const baseContext = {
    ...genContextStitchTrail(),
    ...genContextLogTrail(),
  };
  const context = { ...baseContext, ...getContextOpenAI() };

  // takes a while; skip by default
  given.runIf(!process.env.CI)(
    'a review loop that proposes, reviews, summarizes, and judges code improvement until blockers are gone or 3 attempts max',
    () => {
      const stitcherCodeReviewConcluder = new StitchStepCompute<
        GStitcher<
          Threads<{ critic: Empty }, 'multiple'>,
          typeof context,
          { blockers: string[]; summary: string[] }
        >
      >({
        slug: '[critic]<review:concluder>',
        readme: null,
        form: 'COMPUTE',
        stitchee: 'critic',
        invoke: ({ threads }) => {
          const summary = threads.critic.peers
            .map((peer) => peer.stitches.slice(-1)[0]?.output)
            .filter(Boolean) as string[];
          const blockers = summary.filter((x) => x.includes('blocker'));
          return { input: summary, output: { blockers, summary } };
        },
      });

      const stitcherCodeReviewFanout = asStitcher(
        genStitchFanout({
          slug: '[critic]<code:review>.<fanout>[[review]]',
          readme: null,
          parallels: [
            genStitcherCodeReview({ scope: 'technical', focus: 'blockers' }),
            // genStitcherCodeReview({ scope: 'technical', focus: 'chances' }),
            // genStitcherCodeReview({ scope: 'technical', focus: 'praises' }),
            genStitcherCodeReview({ scope: 'functional', focus: 'blockers' }),
            // genStitcherCodeReview({ scope: 'functional', focus: 'chances' }),
            // genStitcherCodeReview({ scope: 'functional', focus: 'praises' }),
          ],
          concluder: stitcherCodeReviewConcluder,
        }),
      );

      const directorSummarize = new StitchStepCompute<
        GStitcher<
          Threads<{ director: Empty; critic: Empty }>,
          typeof context,
          { directive: string; blockers: string[] }
        >
      >({
        slug: '[director]<summarize>',
        form: 'COMPUTE',
        stitchee: 'director',
        readme: 'turn critic summary into a director directive',
        invoke: ({ threads }) => {
          const last = threads.critic.stitches.slice(-1)[0]?.output;
          return {
            input: last,
            output: {
              directive: `resolve blockers, blockers: ${JSON.stringify(
                last?.blockers,
                null,
                2,
              )}`,
              blockers: last?.blockers,
            },
          };
        },
      });

      const artistCodeProposeRoute = asStitcher(
        genStitchRoute({
          slug: '[artist]<code:propose>',
          readme: 'imagine diff, then write to file',
          sequence: [
            stitcherCodeDiffImagine,
            genStitcherCodeFileWrite<'artist', Threads<{ artist: Empty }>>({
              stitchee: 'artist',
            }),
          ],
        }),
      );

      const criticCodeReviewRoute = asStitcher(
        genStitchRoute({
          slug: '[critic]<code:review>',
          readme: 'review the code from multiple perspectives',
          sequence: [
            genStitcherCodeFileRead<
              'critic',
              Threads<{ artist: Empty; critic: Empty }>
            >({
              stitchee: 'critic',
              output: ({ threads }) =>
                (threads.artist?.stitches.slice(-1)[0]?.output as any) ??
                UnexpectedCodePathError.throw(
                  'expected to find file write output stitch',
                  { threads },
                ),
            }),
            stitcherCodeReviewFanout,
          ],
        }),
      );

      const codeIterateRoute = asStitcherFlat<
        GStitcher<
          Threads<{
            artist: { tools: string[]; facts: string[] };
            critic: { tools: string[]; facts: string[] };
            director: Empty;
          }>,
          ContextOpenAI & GStitcher['context']
        >
      >(
        genStitchRoute({
          slug: '[code:iterate]',
          readme: 'one pass of propose + review + summarize',
          sequence: [
            artistCodeProposeRoute,
            criticCodeReviewRoute,
            directorSummarize,
          ],
        }),
      );

      const judgeDeciderNextStep = new StitchStepCompute<
        GStitcher<
          Threads<{ critic: Empty; judge: Empty }>,
          typeof context,
          { choice: 'release' | 'repeat' }
        >
      >({
        slug: '[judge]<decide:release?>',
        form: 'COMPUTE',
        stitchee: 'judge',
        readme: 'release if more than 3 reviews',
        invoke: ({ threads }) => {
          const choice =
            threads.critic.stitches.length >= 3 ? 'release' : 'repeat';
          return {
            input: { stitches: threads.critic.stitches.length },
            output: { choice },
          };
        },
      });

      const mechanicCodeDiffIterateUntilRelease = asStitcher(
        genStitchCycle({
          slug: '[mechanic]<code:iterate-until-release>',
          readme: 'iterate until no blockers or max 3 tries',
          repeatee: codeIterateRoute,
          decider: judgeDeciderNextStep,
        }),
      );

      const stitcherRestitchLatestArtistOutput = new StitchStepCompute<
        GStitcher<
          Threads<{ artist: Empty }>,
          typeof context,
          { restitched: string }
        >
      >({
        slug: '[artist]<restitch:latest>',
        readme: 'Restitch artists latest output for final record',
        form: 'COMPUTE',
        stitchee: 'artist',
        invoke: ({ threads }) => {
          const latest = threads.artist.stitches.slice(-1)[0];
          if (!latest?.output || typeof latest.output !== 'object')
            throw new Error('No valid artist output to restitch');
          return latest;
        },
      });

      const rootRouteCodeDiff = genStitchRoute({
        slug: '[mechanic]<code:diff>',
        readme:
          'read initial file, then iterate against directive until no blockers or 3 tries',
        sequence: [
          genStitcherCodeFileRead<'artist', Threads<{ artist: Empty }>>({
            stitchee: 'artist',
            output: {
              path: '__path__',
              content: `
/**
 * .what = calls the open-meteo weather api
 * .how =
 *   - uses procedural pattern
 *   - fails fast
 */
export const sdkOpenMeteo = {
  getWeather: (input: {...}, context: VisualogicContext & AuthContextOpenMeteo) => {
    ...
  }
}
          `.trim(),
            },
          }),
          mechanicCodeDiffIterateUntilRelease,
          stitcherRestitchLatestArtistOutput,
        ] as const,
      });
      console.log(JSON.stringify(rootRouteCodeDiff, null, 2));

      when(
        'executing the full review loop with realistic threads for artist, critic, director, and judge',
        () => {
          let threads: Threads<{
            artist: { tools: string[]; facts: string[] };
            critic: { tools: string[]; facts: string[] };
            director: Empty;
            judge: Empty;
          }>;

          beforeAll(async () => {
            threads = {
              artist: await getExampleThreadCodeArtist(),
              critic: await getExampleThreadCodeCritic(),
              director: exampleThreadDirector.clone({
                stitches: [
                  {
                    uuid: getUuid(),
                    createdAt: asUniDateTime(new Date()),
                    input: null,
                    output: { directive: 'fillout the stubout' },
                    stitcher: null,
                    trail: null,
                  }, // start with this directive
                ],
              }),
              judge: genThread({ role: 'judge' }),
            };
          });

          then('it should have a readable route plan declared', () => {
            // expect(rootRouteCodeDiff).toMatchSnapshot();
            expect(
              JSON.stringify(rootRouteCodeDiff, null, 2),
            ).toMatchSnapshot();
          });

          then(
            'it should exit after 3 cycles if blockers are always returned',
            async () => {
              const output = await enweaveOneStitcher(
                { stitcher: rootRouteCodeDiff, threads },
                context,
              );

              expect(output.threads.artist.stitches.length).toBeGreaterThan(0);
              expect(output.threads.critic.stitches.length).toBeGreaterThan(0);
              expect(output.threads.judge.stitches.length).toBeGreaterThan(0);

              expect(
                JSON.stringify(
                  deepReplaceShortUuidsWithLetters(
                    deepOmit(output, [
                      'occurredAt',
                      'createdAt',
                      'input',
                      'output',
                      'tools',
                      'facts',
                      'uuid',
                      'stitchUuid',
                    ]),
                  ),
                  null,
                  2,
                ),
              ).toMatchSnapshot();
            },
          );
        },
      );
    },
  );
});

export function deepOmit(obj: unknown, keys: string[]): unknown {
  if (Array.isArray(obj)) {
    return obj.map((item) => deepOmit(item, keys));
  }

  if (obj !== null && typeof obj === 'object') {
    const raw = { ...obj };
    for (const key of keys) {
      delete (raw as any)[key];
    }

    return Object.fromEntries(
      Object.entries(raw).map(([k, v]) => [k, deepOmit(v, keys)]),
    );
  }

  return obj;
}

const shortUuidRegex = /\(([0-9a-f]{7})\)/gi;

function getLetterLabel(index: number): string {
  let label = '';
  while (index >= 0) {
    label = String.fromCharCode((index % 26) + 65) + label;
    index = Math.floor(index / 26) - 1;
  }
  return label;
}

export function deepReplaceShortUuidsWithLetters(obj: unknown): unknown {
  const seen = new Map<string, string>();
  let index = 0;

  const replace = (value: unknown): unknown => {
    if (typeof value === 'string') {
      return value.replace(shortUuidRegex, (_, shortId) => {
        const key = shortId.toLowerCase();
        if (!seen.has(key)) {
          seen.set(key, getLetterLabel(index++));
        }
        return `(${seen.get(key)!})`;
      });
    }

    if (Array.isArray(value)) {
      return value.map(replace);
    }

    if (value !== null && typeof value === 'object') {
      return Object.fromEntries(
        Object.entries(value).map(([k, v]) => [k, replace(v)]),
      );
    }

    return value;
  };

  return replace(obj);
}
