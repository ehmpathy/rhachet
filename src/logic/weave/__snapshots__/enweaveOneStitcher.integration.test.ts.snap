// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enweaveOneStitcher given: a review loop that proposes, reviews, summarizes, and judges code improvement until blockers are gone or 3 attempts max when: executing the full review loop with realistic threads for artist, critic, director, and judge then: it should exit after 3 cycles if blockers are always returned 1`] = `
"{
  "stitch": {
    "stitcher": {
      "form": "COMPUTE",
      "slug": "[artist]<restitch:latest>",
      "readme": "Restitch artists latest output for final record"
    },
    "trail": {
      "desc": "@[mechanic]<code:diff>(A) > @[artist]<restitch:latest>(B)"
    }
  },
  "threads": {
    "artist": {
      "context": {
        "role": "artist"
      },
      "stitches": [
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[any]<code:file.read>",
            "readme": null
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[any]<code:file.read>(C)"
          }
        },
        {
          "stitcher": {
            "form": "IMAGINE",
            "slug": "[artist]<code:diff>",
            "readme": "intent(imagines a codediff)"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](E) > @[artist]<code:propose>(F) > @[artist]<code:diff>(G)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[any]<code:file.write>",
            "readme": "mock file writer that copies the output string into file output"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](E) > @[artist]<code:propose>(F) > @[any]<code:file.write>(H)"
          }
        },
        {
          "stitcher": {
            "form": "IMAGINE",
            "slug": "[artist]<code:diff>",
            "readme": "intent(imagines a codediff)"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](I) > @[artist]<code:propose>(J) > @[artist]<code:diff>(K)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[any]<code:file.write>",
            "readme": "mock file writer that copies the output string into file output"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](I) > @[artist]<code:propose>(J) > @[any]<code:file.write>(L)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[artist]<restitch:latest>",
            "readme": "Restitch artists latest output for final record"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[artist]<restitch:latest>(B)"
          }
        }
      ]
    },
    "critic": {
      "context": {
        "role": "critic"
      },
      "stitches": [
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[any]<code:file.read>",
            "readme": null
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](E) > @[critic]<code:review>(M) > @[any]<code:file.read>(N)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[critic]<review:concluder>",
            "readme": null
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](E) > @[critic]<code:review>(M) > @[critic]<code:review>.<fanout>[[review]](O) > @[critic]<review:concluder>(P)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[any]<code:file.read>",
            "readme": null
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](I) > @[critic]<code:review>(Q) > @[any]<code:file.read>(R)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[critic]<review:concluder>",
            "readme": null
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](I) > @[critic]<code:review>(Q) > @[critic]<code:review>.<fanout>[[review]](S) > @[critic]<review:concluder>(T)"
          }
        }
      ]
    },
    "director": {
      "context": {
        "role": "director"
      },
      "stitches": [
        {
          "stitcher": null,
          "trail": null
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[director]<summarize>",
            "readme": "turn critic summary into a director directive"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](E) > @[director]<summarize>(U)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[director]<summarize>",
            "readme": "turn critic summary into a director directive"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[code:iterate](I) > @[director]<summarize>(V)"
          }
        }
      ]
    },
    "judge": {
      "context": {
        "role": "judge"
      },
      "stitches": [
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[judge]<decide:release?>",
            "readme": "release if more than 3 reviews"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[judge]<decide:release?>(W)"
          }
        },
        {
          "stitcher": {
            "form": "COMPUTE",
            "slug": "[judge]<decide:release?>",
            "readme": "release if more than 3 reviews"
          },
          "trail": {
            "desc": "@[mechanic]<code:diff>(A) > @[mechanic]<code:iterate-until-release>(D) > @[judge]<decide:release?>(X)"
          }
        }
      ]
    }
  }
}"
`;

exports[`enweaveOneStitcher given: a review loop that proposes, reviews, summarizes, and judges code improvement until blockers are gone or 3 attempts max when: executing the full review loop with realistic threads for artist, critic, director, and judge then: it should have a readable route plan declared 1`] = `
"{
  "form": "ROUTE",
  "slug": "[mechanic]<code:diff>",
  "readme": "read initial file, then iterate against directive until no blockers or 3 tries",
  "sequence": [
    {
      "form": "COMPUTE",
      "readme": null,
      "slug": "[any]<code:file.read>",
      "stitchee": "artist"
    },
    {
      "form": "CYCLE",
      "slug": "[mechanic]<code:iterate-until-release>",
      "readme": "iterate until no blockers or max 3 tries",
      "repeatee": {
        "form": "ROUTE",
        "slug": "[code:iterate]",
        "readme": "one pass of propose + review + summarize",
        "sequence": [
          {
            "form": "ROUTE",
            "slug": "[artist]<code:propose>",
            "readme": "imagine diff, then write to file",
            "sequence": [
              {
                "form": "IMAGINE",
                "slug": "[artist]<code:diff>",
                "stitchee": "artist",
                "readme": "intent(imagines a codediff)"
              },
              {
                "form": "COMPUTE",
                "slug": "[any]<code:file.write>",
                "readme": "mock file writer that copies the output string into file output",
                "stitchee": "artist"
              }
            ]
          },
          {
            "form": "ROUTE",
            "slug": "[critic]<code:review>",
            "readme": "review the code from multiple perspectives",
            "sequence": [
              {
                "form": "COMPUTE",
                "readme": null,
                "slug": "[any]<code:file.read>",
                "stitchee": "critic"
              },
              {
                "form": "FANOUT",
                "slug": "[critic]<code:review>.<fanout>[[review]]",
                "readme": null,
                "parallels": [
                  {
                    "form": "IMAGINE",
                    "slug": "[critic]:<code:review:technical:blockers>",
                    "stitchee": "critic",
                    "readme": "intent(imagines constructive feedback over the latest state of code, focused on technical blockers)"
                  },
                  {
                    "form": "IMAGINE",
                    "slug": "[critic]:<code:review:functional:blockers>",
                    "stitchee": "critic",
                    "readme": "intent(imagines constructive feedback over the latest state of code, focused on functional blockers)"
                  }
                ],
                "concluder": {
                  "slug": "[critic]<review:concluder>",
                  "readme": null,
                  "form": "COMPUTE",
                  "stitchee": "critic"
                }
              }
            ]
          },
          {
            "slug": "[director]<summarize>",
            "form": "COMPUTE",
            "stitchee": "director",
            "readme": "turn critic summary into a director directive"
          }
        ]
      },
      "decider": {
        "slug": "[judge]<decide:release?>",
        "form": "COMPUTE",
        "stitchee": "judge",
        "readme": "release if more than 3 reviews"
      },
      "halter": {
        "threshold": {
          "repetitions": 100,
          "duration": {
            "hours": 1
          }
        }
      }
    },
    {
      "slug": "[artist]<restitch:latest>",
      "readme": "Restitch artists latest output for final record",
      "form": "COMPUTE",
      "stitchee": "artist"
    }
  ]
}"
`;
